pkg/storage/image.go

```diff
@@ -0,0 +1,207 @@
+			results = append(results, ImageResult{
+				ID:    image.ID,
+				Names: image.Names,
+			})
+		}
+	} else {
+		images, err := svc.store.Images()
+		if err != nil {
+			return nil, err
+		}
+		for _, image := range images {
+			results = append(results, ImageResult{
+				ID:    image.ID,
+				Names: image.Names,
+			})
+		}
+	}
+	return results, nil
+}
+
+func (svc *imageService) ImageStatus(nameOrID string) (*ImageResult, error) {
+	ref, err := transports.ParseImageName(nameOrID)
+	if err != nil {
+		ref2, err2 := istorage.Transport.ParseStoreReference(svc.store, "@"+nameOrID)
+		if err2 != nil {
+			ref3, err3 := istorage.Transport.ParseStoreReference(svc.store, nameOrID)
+			if err3 != nil {
+				return nil, err
+			}
+			ref2 = ref3
+		}
+		ref = ref2
+	}
+	image, err := istorage.Transport.GetStoreImage(svc.store, ref)
+	if err != nil {
+		return nil, err
+	}
+
+	systemContext := types.SystemContext{}
+	img, err := ref.NewImage(&systemContext)
+	if err != nil {
+		return nil, err
+	}
+	size := imageSize(img)
+	img.Close()
+
+	return &ImageResult{
+		ID:    image.ID,
+		Names: image.Names,
+		Size:  size,
+	}, nil
+}
+
+func imageSize(img types.Image) *uint64 {
+	if sizer, ok := img.(istorage.Sizer); ok {
+		sum, err := sizer.Size()
+		if err != nil || sum < 0 {
+			return nil
+		}
+		usum := uint64(sum)
+		return &usum
+	}
+	return nil
+}
+
+func (svc *imageService) PullImageUsingContexts(imageName string, policyContext *signature.PolicyContext, options *copy.Options) (types.ImageReference, error) {
+	if imageName == "" {
+		return nil, storage.ErrNotAnImage
+	}
+	srcRef, err := transports.ParseImageName(imageName)
+	if err != nil {
+		srcRef2, err2 := transports.ParseImageName(svc.defaultTransport + imageName)
+		if err2 != nil {
+			return nil, err
+		}
+		srcRef = srcRef2
+	}
+	dest := imageName
+	if srcRef.DockerReference() != nil {
+		dest = srcRef.DockerReference().FullName()
+	}
+	destRef, err := istorage.Transport.ParseStoreReference(svc.store, dest)
+	if err != nil {
+		return nil, err
+	}
+	err = copy.Image(policyContext, destRef, srcRef, options)
+	if err != nil {
+		return nil, err
+	}
+	// Go find the image, and attach the requested name to it, so that we
+	// can more easily find it later, even if the destination reference
+	// looks different.
+	destImage, err := istorage.Transport.GetStoreImage(svc.store, destRef)
+	if err != nil {
+		return nil, err
+	}
+	names := append(destImage.Names, imageName, dest)
+	err = svc.store.SetNames(destImage.ID, names)
+	if err != nil {
+		return nil, err
+	}
+	return destRef, nil
+}
+
+func (svc *imageService) PullImage(imageName string) (types.ImageReference, error) {
+	systemContext := types.SystemContext{}
+	policy, err := signature.DefaultPolicy(&systemContext)
+	if err != nil {
+		return nil, err
+	}
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return nil, err
+	}
+	options := copy.Options{}
```

Comment:
Probably need to provide a way to pass in the options structure directly.  Adding that.